verdict/core/pipeline.py

```diff
@@ -21,29 +21,58 @@
 from verdict.util.exceptions import VerdictDeclarationTimeError
 from verdict.util.log import init_logger, logger
 from verdict.util.misc import keyboard_interrupt_safe
+from verdict.util.tracing import Tracer, NoOpTracer, TracingManager
+import uuid
+
 
 
 class Pipeline:
     name: str
     block: Block
 
     executor: GraphExecutor
+    tracer: Tracer
+
+    def __init__(self, name: str = 'Pipeline', tracer: Optional[Tracer] = None) -> None:
+        """
+        Initialize a Pipeline.
 
-    def __init__(self, name: str='Pipeline') -> None:
+        Args:
+            name: The name of the pipeline.
+            tracer: The tracer or tracing manager to use. If None, uses NoOpTracer.
+        """
         super().__init__()
         self.name = name
         self.block = Block()
+        if tracer is None:
+            self.tracer = TracingManager([NoOpTracer()])
+        elif isinstance(tracer, TracingManager):
+            self.tracer = tracer
+        else:
+            self.tracer = TracingManager([tracer])
+
+    def add_tracer(self, tracer: Tracer) -> None:
```

Comment:
Unused?

Reply:
The `add_tracer` is to allow user's to add their own tracers at runtime, after the creation of a pipeline or before using a `pipeline.run`